File: VideoMergeAndroid/app/src/main/java/com/moon/videomerger/merger/CollageMerger.kt
package com.moon.videomerger.merger

import com.moon.videomerger.util.VideoMeta

/**
 * Collage 模式合并器 —— 一个主窗口 + 若干副窗口，严格无空隙铺满画布。
 *
 * 核心逻辑移植自 collage_merge.py，包含 logo 截断，简化掉人脸检测。
 * cover 模式下使用正中裁剪（人脸检测需要 OpenCV，在 Android 上可选集成）。
 */
class CollageMerger {

    fun buildCommand(
        inputPaths: List<String>,
        durations: List<Double>,
        metas: List<VideoMeta>,
        outputPath: String,
        options: MergeOptions,
        cutTimes: List<Double?> = emptyList()
    ): String {
        val n = inputPaths.size
        val mainIdx = options.collageMainIndex.coerceIn(0, n - 1)

        val canvasW = options.canvasWidth.toAligned16()
        val canvasH = options.canvasHeight.toAligned16()

        // 与 UI 预览共用 MergeLayout，保证所见即所得
        val layout = MergeLayout.collageLayout(
            n, canvasW, canvasH, options.collageMainRatio,
            options.collageOrient, options.gap, mainIdx
        )
        val sizes = layout.sizes
        val positions = layout.positions

        val maxDur = durations.maxOrNull() ?: 0.0
        val audioMask = metas.map { it.hasAudio }
        val faceXY = List(n) { 0.5 to 0.5 }  // 正中裁剪

        val (filterComplex, hasAudio) = buildFilter(
            n, sizes, positions, durations, maxDur, audioMask,
            options.collageFit, faceXY, cutTimes
        )

        val cmd = StringBuilder().apply {
            append("-y ")
            inputPaths.forEach { append("-i \"$it\" ") }
            append("-filter_complex \"$filterComplex\" ")
            append("-map \"[vout]\" ")
            append("-t ${maxDur.fmt()} ")
            append("-c:v h264_mediacodec -b:v 8M ")
            if (hasAudio) {
                append("-map \"[aout]\" -c:a aac -b:a 192k ")
            }
            append("\"$outputPath\"")
        }

        return cmd.toString()
    }


    // ---------- filter 构建 ----------

    private fun scaleFill(cw: Int, ch: Int, fit: String, fx: Double = 0.5, fy: Double = 0.5): String {
        return if (fit == "contain") {
            "scale=${cw}:${ch}:force_original_aspect_ratio=decrease," +
            "pad=${cw}:${ch}:(ow-iw)/2:(oh-ih)/2:black,fps=30,setsar=1,format=yuv420p"
        } else {
            // cover: crop 偏移
            "scale=${cw}:${ch}:force_original_aspect_ratio=increase," +
            "crop=${cw}:${ch}:${fx.fmt(4)}*(in_w-out_w):${fy.fmt(4)}*(in_h-out_h),fps=30,setsar=1,format=yuv420p"
        }
    }

    private fun buildFilter(
        n: Int,
        sizes: List<Pair<Int, Int>>,
        positions: List<Pair<Int, Int>>,
        durations: List<Double>,
        maxDur: Double,
        audioMask: List<Boolean>,
        fit: String,
        faceXY: List<Pair<Double, Double>>,
        cutTimes: List<Double?>
    ): Pair<String, Boolean> {
        val scaled = mutableListOf<String>()

        for (i in 0 until n) {
            val (cw, ch) = sizes[i]
            val (fx, fy) = faceXY[i]
            val padDur = (maxDur - durations[i]).coerceAtLeast(0.0)
            val trimPfx = if (i < cutTimes.size && cutTimes[i] != null) "trim=end=${cutTimes[i]!!.fmt(3)},setpts=PTS-STARTPTS," else ""
            val common = scaleFill(cw, ch, fit, fx, fy) + "[v$i]"

            if (padDur <= 0.001) {
                scaled.add("[$i:v]$trimPfx$common")
            } else {
                scaled.add(
                    "[$i:v]$trimPfx" + "split=2[${i}A][${i}B];" +
                    "[${i}A]trim=end_frame=1,loop=loop=-1:size=1,setpts=PTS-STARTPTS[${i}Fof];" +
                    "[${i}B]setpts=PTS-STARTPTS[${i}M];" +
                    "[${i}M][${i}Fof]concat=n=2:v=1:a=0[${i}C];" +
                    "[${i}C]trim=end=${maxDur.fmt()},setpts=PTS-STARTPTS,$common"
                )
            }
        }

        val layouts = positions.map { "${it.first}_${it.second}" }
        val inputsConcat = (0 until n).joinToString("") { "[v$it]" }
        val layoutStr = layouts.joinToString("|")

        val parts = mutableListOf<String>()
        parts.add(scaled.joinToString(";"))
        parts.add("$inputsConcat xstack=inputs=$n:layout=$layoutStr:fill=black[vout]")

        // 音频
        val audioParts = mutableListOf<String>()
        for (i in 0 until n) {
            if (!audioMask[i]) continue
            val padA = (maxDur - durations[i]).coerceAtLeast(0.0)
            val aTrim = if (i < cutTimes.size && cutTimes[i] != null) "atrim=end=${cutTimes[i]!!.fmt(3)},asetpts=PTS-STARTPTS," else ""
            val apad = if (padA > 0) ",apad=whole_dur=${maxDur.fmt()}" else ""
            audioParts.add("[$i:a]${aTrim}aresample=44100$apad[a$i]")
        }

        if (audioParts.isNotEmpty()) {
            val amixIn = audioMask.indices.filter { audioMask[it] }.joinToString("") { "[a$it]" }
            parts.add(audioParts.joinToString(";"))
            parts.add("$amixIn amix=inputs=${audioParts.size}:duration=first:normalize=0[aout]")
        }

        return parts.joinToString(";") to audioParts.isNotEmpty()
    }
}

File: VideoMergeIOS/VideoMerger/Merger/CollageMerger.swift
//
//  CollageMerger.swift
//  VideoMerger
//
//  iOS 移植自 Android VideoMerger 的 CollageMerger.kt
//  算法源自 collage_merge.py（简化掉人脸检测，使用正中裁剪）
//

import Foundation

/// Collage 模式合并器 —— 一个主窗口 + 若干副窗口，严格无空隙铺满画布
final class CollageMerger {

    func buildCommand(inputPaths: [String],
                      durations: [Double],
                      metas: [VideoMeta],
                      outputPath: String,
                      options: MergeOptions,
                      cutTimes: [Double?]) -> String {
        let n = inputPaths.count
        let mainIdx = max(0, min(options.collageMainIndex, n - 1))

        let canvasW = options.canvasWidth.toAligned16
        let canvasH = options.canvasHeight.toAligned16

        // 与 UI 预览共用 MergeLayout，保证所见即所得
        let layout = MergeLayout.collageLayout(
            n: n, canvasW: canvasW, canvasH: canvasH,
            mainRatio: options.collageMainRatio,
            orient: options.collageOrient.rawValue,
            gap: options.gap, mainIdx: mainIdx
        )
        let sizes = layout.sizes
        let positions = layout.positions

        let maxDur = durations.max() ?? 0.0
        let audioMask = metas.map { $0.hasAudio }
        let faceXY = Array(repeating: (0.5, 0.5), count: n)

        let (filterComplex, hasAudio) = buildFilter(
            n: n, sizes: sizes, positions: positions,
            durations: durations, maxDur: maxDur, audioMask: audioMask,
            fit: options.collageFit.rawValue, faceXY: faceXY, cutTimes: cutTimes
        )

        var cmd = "-y "
        for p in inputPaths { cmd += "-i \"\(p)\" " }
        cmd += "-filter_complex \"\(filterComplex)\" "
        cmd += "-map \"[vout]\" "
        cmd += "-t \(maxDur.fmt()) "
        cmd += "-c:v h264_videotoolbox -b:v 8M -pix_fmt yuv420p " // ★ 硬件编码（对齐 Android mediacodec 与编辑器 videotoolbox）
        if hasAudio {
            cmd += "-map \"[aout]\" -c:a aac -b:a 192k "
        }
        cmd += "\"\(outputPath)\""
        return cmd
    }

    // MARK: - filter

    private func scaleFill(cw: Int, ch: Int, fit: String, fx: Double = 0.5, fy: Double = 0.5) -> String {
        if fit == "contain" {
            return "scale=\(cw):\(ch):force_original_aspect_ratio=decrease," +
                   "pad=\(cw):\(ch):(ow-iw)/2:(oh-ih)/2:black,fps=30,setsar=1,format=yuv420p"
        } else {
            // cover: crop 偏移
            return "scale=\(cw):\(ch):force_original_aspect_ratio=increase," +
                   "crop=\(cw):\(ch):\(fx.fmt(4))*(in_w-out_w):\(fy.fmt(4))*(in_h-out_h),fps=30,setsar=1,format=yuv420p"
        }
    }

    private func buildFilter(n: Int,
                             sizes: [(Int, Int)],
                             positions: [(Int, Int)],
                             durations: [Double],
                             maxDur: Double,
                             audioMask: [Bool],
                             fit: String,
                             faceXY: [(Double, Double)],
                             cutTimes: [Double?]) -> (String, Bool) {
        var scaled: [String] = []

        for i in 0..<n {
            let (cw, ch) = sizes[i]
            let (fx, fy) = faceXY[i]
            let padDur = max(0.0, maxDur - durations[i])
            let trimPfx: String = (i < cutTimes.count && cutTimes[i] != nil)
                ? "trim=end=\(cutTimes[i]!.fmt(3)),setpts=PTS-STARTPTS,"
                : ""
            let common = scaleFill(cw: cw, ch: ch, fit: fit, fx: fx, fy: fy) + "[v\(i)]"

            if padDur <= 0.001 {
                scaled.append("[\(i):v]\(trimPfx)\(common)")
            } else {
                scaled.append(
                    "[\(i):v]\(trimPfx)" +
                    "split=2[\(i)A][\(i)B];" +
                    "[\(i)A]trim=end_frame=1,loop=loop=-1:size=1,setpts=PTS-STARTPTS[\(i)Fof];" +
                    "[\(i)B]setpts=PTS-STARTPTS[\(i)M];" +
                    "[\(i)M][\(i)Fof]concat=n=2:v=1:a=0[\(i)C];" +
                    "[\(i)C]trim=end=\(maxDur.fmt()),setpts=PTS-STARTPTS,\(common)"
                )
            }
        }

        let layouts = positions.map { "\($0.0)_\($0.1)" }
        let inputsConcat = (0..<n).map { "[v\($0)]" }.joined()
        let layoutStr = layouts.joined(separator: "|")

        var parts: [String] = []
        parts.append(scaled.joined(separator: ";"))
        parts.append("\(inputsConcat) xstack=inputs=\(n):layout=\(layoutStr):fill=black[vout]")

        // 音频
        var audioParts: [String] = []
        for i in 0..<n {
            if !audioMask[i] { continue }
            let padA = max(0.0, maxDur - durations[i])
            let aTrim: String = (i < cutTimes.count && cutTimes[i] != nil)
                ? "atrim=end=\(cutTimes[i]!.fmt(3)),asetpts=PTS-STARTPTS,"
                : ""
            let apad = padA > 0 ? ",apad=whole_dur=\(maxDur.fmt())" : ""
            audioParts.append("[\(i):a]\(aTrim)aresample=44100\(apad)[a\(i)]")
        }

        if !audioParts.isEmpty {
            let amixIn = (0..<n).filter { audioMask[$0] }.map { "[a\($0)]" }.joined()
            parts.append(audioParts.joined(separator: ";"))
            parts.append("\(amixIn) amix=inputs=\(audioParts.count):duration=first:normalize=0[aout]")
        }

        return (parts.joined(separator: ";"), !audioParts.isEmpty)
    }
}
